contracts/IdleYieldSource.sol

```diff
@@ -80,38 +94,55 @@ contract IdleYieldSource is IProtocolYieldSource, Initializable, ReentrancyGuard
+
+    /// @notice Calculates the current price per share
+    /// @return Average idleToken price for this contract
+    function _price() internal view returns (uint256) {
+      return idleToken.tokenPriceWithFee(address(this));
     }
 
     /// @notice Calculates the number of shares that should be mint or burned when a user deposit or withdraw
     /// @param tokens Amount of tokens
-    /// return Number of shares
-    function _tokenToShares(uint256 tokens) internal view returns (uint256 shares) {
-        shares = (tokens * ONE_IDLE_TOKEN) / _price();
+    /// @return Number of shares
+    function _tokenToShares(uint256 tokens) internal view returns (uint256) {
+        uint256 shares = 0;
+        uint256 totalSupply = totalSupply();
+
+        if (totalSupply == 0) {
+            shares = tokens;
+        } else {
+            // rate = tokens / shares
+            // shares = tokens * (totalShares / yieldSourceTotalSupply)
+            uint256 exchangeMantissa = FixedPoint.calculateMantissa(totalSupply, idleToken.balanceOf(address(this)).mul(_price()).div(ONE_IDLE_TOKEN));
+            shares = FixedPoint.multiplyUintByMantissa(tokens, exchangeMantissa);
+        }
+
+        return shares;
     }
 
     /// @notice Calculates the number of tokens a user has in the yield source
     /// @param shares Amount of shares
-    /// return Number of tokens
-    function _sharesToToken(uint256 shares) internal view returns (uint256 tokens) { 
-        tokens = (shares * _price()) / ONE_IDLE_TOKEN;
-    }
-
-    /// @notice Calculates the current price per share
-    /// @return avg idleToken price for this contract
-    function _price() internal view returns (uint256) {
-      return IIdleToken(idleToken).tokenPriceWithFee(address(this));
+    /// @return Number of tokens
+    function _sharesToToken(uint256 shares) internal view returns (uint256) {
+        uint256 tokens = 0;
+
+        if (totalSupply() == 0) {
+            tokens = shares;
+        } else {
+            // tokens = shares * (yieldSourceTotalSupply / totalShares)
+            uint256 exchangeMantissa = FixedPoint.calculateMantissa(idleToken.balanceOf(address(this)).mul(_price()).div(ONE_IDLE_TOKEN), totalSupply());
+            tokens = FixedPoint.multiplyUintByMantissa(shares, exchangeMantissa);
+        }
+
+        return tokens;
     }
 
     /// @notice Deposit asset tokens to Idle
     /// @param mintAmount The amount of asset tokens to be deposited
     /// @return number of minted tokens
     function _depositToIdle(uint256 mintAmount) internal returns (uint256) {
-        IERC20Upgradeable(underlyingAsset).safeTransferFrom(msg.sender, address(this), mintAmount);
-        return IIdleToken(idleToken).mintIdleToken(mintAmount, false, address(0));
+        IERC20Upgradeable(_tokenAddress()).safeTransferFrom(msg.sender, address(this), mintAmount);
+        return idleToken.mintIdleToken(mintAmount, false, address(0));
```

Comment:
interesting...do we want to set the referral address here?  Maybe make it ownable and allow owner to set a referrer?

Reply:
Should we set a default referral address or should it return address zero by default?

Reply:
I think we should set one in the constructor

Reply:
Which address should we use then?

Reply:
Done in this commit: https://github.com/pooltogether/idle-yield-source/pull/5/commits/d7209fd106365a9553aaf070fa22d532ab756cd4

Should we also initialize the yield source by using the generic proxy factory?

Also, I'm not sure we should set the param `_skipWholeRebalance` of the `mintidletoken` function, to `false` all the time. May be pretty costly in gas for small deposits.
https://developers.idle.finance/methods/mintidletoken